### PR TITLE
Add external interop

### PR DESCRIPTION
**Pull request by matyas-streamhpc**

### CODE REVIEW

File: docs/how-to/hip_runtime_api/external_interop.rst [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+creating a Vulkan instance, and configuring the GPU device and queue. After
+these initial steps, the kernel executes the sine wave computation, and Vulkan
+continuously updates the window framebuffer to display the computed data
+until the window is closed.
+
+The Vulkan external descriptor is defined in the following way:
+
+.. literalinclude:: ../../tools/example_codes/external_interop.hip
+   :start-after: // [Sphinx external descriptor start]
+   :end-before: // [Sphinx external descriptor end]
```
Comment (neon60) [inline docs/how-to/hip_runtime_api/external_interop.rst:65]: Please add ```:language: cpp```

File: docs/how-to/hip_runtime_api/external_interop.rst [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+The Vulkan semaphore is converted to HIP semaphore shown in the following example:
+
+.. literalinclude:: ../../tools/example_codes/external_interop.hip
+   :start-after: // [Sphinx semaphore use in kernel start]
+   :end-before: // [Sphinx semaphore use in kernel end]
+
```
Comment (neon60) [inline docs/how-to/hip_runtime_api/external_interop.rst:89]:
```suggestion
    :language: cpp
    
```

File: docs/how-to/hip_runtime_api/external_interop.rst [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+The external memory is used in the following example:
+
+.. literalinclude:: ../../tools/example_codes/external_interop.hip
+   :start-after: // [Sphinx external memory use in kernel start]
+   :end-before: // [Sphinx external memory use in kernel end]
+
```
Comment (neon60) [inline docs/how-to/hip_runtime_api/external_interop.rst:120]:
```suggestion
    :language: cpp
    
```

File: docs/how-to/hip_runtime_api/external_interop.rst [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+these initial steps, the kernel executes the sine wave computation, and Vulkan
+continuously updates the window framebuffer to display the computed data
+until the window is closed.
+
+The Vulkan external descriptor is defined in the following way:
+
+.. literalinclude:: ../../tools/example_codes/external_interop.hip
+   :start-after: // [Sphinx external descriptor start]
+   :end-before: // [Sphinx external descriptor end]
+
```
Comment (neon60) [inline docs/how-to/hip_runtime_api/external_interop.rst:75]:
```suggestion
    :language: cpp
    
```

**Review by neon60 — APPROVED** [review]

File: docs/how-to/hip_runtime_api/external_interop.rst [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+*******************************************************************************
+
+This feature allows HIP to work with resources -- like memory and semaphores --
+created by other APIs. This means resources can be used from APIs like CUDA,
+OpenCL and Vulkan within HIP, making it easier to integrate HIP into existing
+projects.
+
+To use external resources in HIP, you typically follow these steps:
+
+- import the resource: Use functions to import resources from other APIs
```
Comment (randyh62) [inline docs/how-to/hip_runtime_api/external_interop.rst]:
```suggestion
- Import resources from other APIs using HIP provided functions
```

File: docs/how-to/hip_runtime_api/external_interop.rst [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+- use the resource: the imported resources can be used as if they were created
```
Comment (randyh62) [inline docs/how-to/hip_runtime_api/external_interop.rst]:
```suggestion
- Use external resources as if they were created in HIP
```

File: docs/how-to/hip_runtime_api/external_interop.rst [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+created by other APIs. This means resources can be used from APIs like CUDA,
+OpenCL and Vulkan within HIP, making it easier to integrate HIP into existing
+projects.
+
+To use external resources in HIP, you typically follow these steps:
+
+- import the resource: Use functions to import resources from other APIs
+- use the resource: the imported resources can be used as if they were created 
+  by HIP itself
+- destroy the resource: to clean up.
```
Comment (randyh62) [inline docs/how-to/hip_runtime_api/external_interop.rst]:
```suggestion
- Destroy the resource using HIP functions to clean up 
```

File: docs/how-to/hip_runtime_api/external_interop.rst [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+This feature allows HIP to work with resources -- like memory and semaphores --
+created by other APIs. This means resources can be used from APIs like CUDA,
+OpenCL and Vulkan within HIP, making it easier to integrate HIP into existing
+projects.
+
+To use external resources in HIP, you typically follow these steps:
+
+- import the resource: Use functions to import resources from other APIs
+- use the resource: the imported resources can be used as if they were created 
+  by HIP itself
```
Comment (randyh62) [inline docs/how-to/hip_runtime_api/external_interop.rst]:
```suggestion

```

File: docs/how-to/hip_runtime_api/external_interop.rst (hunk 7 above)
Comment (randyh62) [reply docs/how-to/hip_runtime_api/external_interop.rst]: Does this destroy the association in HIP? Or destroy the actual external resource?

File: docs/how-to/hip_runtime_api/external_interop.rst [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+
+Semaphore functions are essential for synchronization in parallel computing.
+These functions facilitate communication and coordination between different
+parts of a program or between different programs. By managing semaphores, tasks
+are executed in the correct order, and resources are utilized effectively.
+Semaphore functions ensure smooth operation, preventing conflicts and
+maintaining the integrity of processes; upholding the integrity and performance
+of concurrent processes.
+
+Semaphore functions are not supported recently on Linux (see: :doc:`../reference/hip_runtime_api/external_interop`.).
```
Comment (randyh62) [inline docs/how-to/hip_runtime_api/external_interop.rst]:
```suggestion
External semaphore functions can be used in HIP as described in  :doc:`../reference/hip_runtime_api/external_interop`.
```

File: docs/how-to/hip_runtime_api/external_interop.rst [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+Memory Functions
+===============================================================================
+
+Memory functions focus on the efficient sharing and management of memory
```
Comment (randyh62) [inline docs/how-to/hip_runtime_api/external_interop.rst]:
```suggestion
HIP external memory functions focus on the efficient sharing and management of memory
```

File: docs/how-to/hip_runtime_api/external_interop.rst [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+resources. These functions allow the importation of memory created by external
```
Comment (randyh62) [inline docs/how-to/hip_runtime_api/external_interop.rst]:
```suggestion
resources. These functions enable importing memory created by external
```

File: docs/how-to/hip_runtime_api/external_interop.rst [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+systems, enabling the program to use this memory seamlessly. Memory functions
```
Comment (randyh62) [inline docs/how-to/hip_runtime_api/external_interop.rst]:
```suggestion
systems, enabling the HIP program to use this memory seamlessly. Memory functions
```

File: docs/how-to/hip_runtime_api/external_interop.rst [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+example. Signaling on the semaphore in one API will allow the other API to wait
+on it, which is how we can guarantee synchronized access to resources in a
+cross-API manner.
+
+.. literalinclude:: ../../tools/example_codes/external_interop.hip
+   :start-after: // [Sphinx semaphore convert start]
+   :end-before: // [Sphinx semaphore convert end]
+   :language: cpp
+
+When the HIP external memory is exported from Vulkan and imported to HIP, it
```
Comment (randyh62) [inline docs/how-to/hip_runtime_api/external_interop.rst]: This is nice detail for the user. 

Is the memory copied into HIP, or is the memory location shared between Vulkan and HIP?

**Review by randyh62 — APPROVED** [review]
Left a few comments.

File: docs/how-to/hip_runtime_api/external_interop.rst (hunk 7 above)
Comment (matyas-streamhpc) [reply docs/how-to/hip_runtime_api/external_interop.rst]: It destroys the HIP object only.

File: docs/how-to/hip_runtime_api/external_interop.rst (hunk 13 above)
Comment (matyas-streamhpc) [reply docs/how-to/hip_runtime_api/external_interop.rst]: The memory is shared (same with semaphores).

The Vulkan API reference, which is used by the example, is located here:
https://docs.vulkan.org/guide/latest/extensions/external.html

Comment (matyas-streamhpc) [reply docs/how-to/hip_runtime_api/external_interop.rst]: Updated

File: docs/how-to/hip_runtime_api/external_interop.rst (hunk 7 above)
Comment (matyas-streamhpc) [reply docs/how-to/hip_runtime_api/external_interop.rst]: Updated

**Review by randyh62 — APPROVED** [review]

**Pull request merged**